server/grpc_service.go

```diff
@@ -1087,3 +1229,51 @@ func (s *Server) validateInternalRequest(header *pdpb.RequestHeader, onlyAllowLe
 	}
 	return nil
 }
+
+func (s *Server) getDelegateClient(ctx context.Context, receiverAddr string) (*grpc.ClientConn, error) {
+	client, ok := s.clientConns.Load(receiverAddr)
+	if !ok {
+		tlsConfig, err := s.GetTLSConfig().ToTLSConfig()
+		if err != nil {
+			return nil, err
+		}
+		cc, err := grpcutil.GetClientConn(ctx, receiverAddr, tlsConfig)
+		if err != nil {
+			return nil, err
+		}
+		client = cc
+		s.clientConns.Store(receiverAddr, cc)
+	}
+	return client.(*grpc.ClientConn), nil
+}
+
+func (s *Server) isLocalRequest(receiverAddr string) bool {
+	if receiverAddr == "" {
+		return true
+	}
+	memberAddrs := s.GetMember().Member().GetClientUrls()
+	for _, addr := range memberAddrs {
+		if addr == receiverAddr {
+			return true
+		}
+	}
+	return false
+}
+
+func (s *Server) createRedirectStream(client *grpc.ClientConn) (pdpb.PD_TsoClient, context.CancelFunc, error) {
+	done := make(chan struct{})
+	ctx, cancel := context.WithCancel(s.ctx)
+	go func(streamCtx context.Context, cancel context.CancelFunc, done chan struct{}) {
+		select {
+		case <-done:
+			return
+		case <-time.After(3 * time.Second):
+			cancel()
+		case <-streamCtx.Done():
+		}
+		<-done
+	}(ctx, cancel, done)
+	redirectStream, err := pdpb.NewPDClient(client).Tso(ctx)
+	done <- struct{}{}
+	return redirectStream, cancel, err
+}
```

Comment:
I think we can record a `TODO` for the case if goroutine here timeout when tso stream created successfully.